### PR TITLE
InfluxDB: Support for visualizing logs in Explore

### LINKED ISSUE #17298: Support InfluxDB for visualizing logs in Explore
Add InfluxDB support for visualizing logs in Explore.

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Inital POC for creating a generic AdHocFilter component in React. First use is for InfluxDB but will be developed more for Elastic in PRs to come.

**Which issue(s) this PR fixes**:
Closes #17298 

**Special notes for your reviewer**:
Has very limited styling and UX.

### CODE REVIEW

File: packages/grafana-ui/src/types/datasource.ts [hunk 1]
```diff
@@ -199,6 +201,10 @@ export abstract class DataSourceApi<
     options?: TContextQueryOptions
   ) => Promise<DataQueryResponse>;
 
+  getTagKeys?(options: any): any[];
+
+  getTagValues?(options: { key: any }): any[];
```
Comment (torkelo) [inline packages/grafana-ui/src/types/datasource.ts]: should be Promise<MetricNameMeta[]> 

which is a new type, not sure what the name should, just a {text: string, value: string} 
same type that metricFindQuery and getTagKeys returns

**Review by torkelo — COMMENTED** [review]
Looks like a great start!

**Comment by davkal** [comment]
High-level feedback:

- great that we got this working!!
- the UI should encourage a simple flow, that focusses more on getting logs to show up (and de-emphasises tag filtering), IMO this is the MVP:
  - the selection of the measurement is the most important step, perhaps hide the tag selector until something has been selected
  - the second most important step is the selection of the value field, there should be a list of value field names where you need to click one, this is more important than selecting tags
  - add a cheat sheet about what to do to get started (select measurement, then select field)
- a stretch goal could focus on filtering
  - I'd prefer it if the adhoc tag filters are the secondary way of filtering. the primary way should be the labels mechanism (currently via the labels column, but in the future possibly via a row expansion) by clicking on the label (tag) itself

**Comment by torkelo** [comment]
Looks like a great start! 

> I'd prefer it if the adhoc tag filters are the secondary way of filtering. the primary way should be the labels mechanism (currently via the labels column, but in the future possibly via a row expansion) by clicking on the label (tag) itself

I agree that being able to filter using the result view is maybe the preferred "drilldown" workflow. But the main problem with that as the first step of narrowing down the result is that I think it won't work.

In Loki you specify a label (pod or app or instance) to have as a first filter and then you can narrow down further using the filtering from the results. We need something similar for InfluxDB. By only specifying measurement and field (which will usually be something like logs & message) you will basically query ALL your logs, so the first 1000 entries form all your log sources. Will just be plain luck if you can find the label (app/instance/etc) in that first result set.

File: packages/grafana-ui/src/types/datasource.ts (hunk 1 above)
Comment (marefr) [reply packages/grafana-ui/src/types/datasource.ts]: Added `metricFindQuery` to `DatasourceAPI` and now uses a `Promise<MetricFindValue>` as returned type. `getTagKeys` and `getTagValues` uses the same returned type. Hard with naming, but I think `MetricFindValue` could work in the first iteration.

**Comment by marefr** [comment]
Tag selector is now hidden until a measurement has been selected

**Comment by marefr** [comment]
@hugohaggmark guess next step is to look into allow user to select which field to use as the log message.  Guess we need to add a field selection to the query field since you can have multiple measurements. The alternative would be to have it as a configuration on datasource but gets tricky if you have multiple measurements with logs.

Sometime soon I think the easiest and most straightforward way is to have some sort of general field selection that works for all datasources and in that case I would find it easier if this selection is made in the logs visualization area compared to in the query field.

**Review by marefr — APPROVED** [review]
LGTM! This is enough to merge to master :+1: Going to create issues of outstanding things

**Comment by jcdauchy-moodys** [comment]
Is there a way to give the "level" column so that the LogPanel can retrieve the loglevel information ?

**Comment by ivanahuckova** [comment]
Hello @jcdauchy-moodys! The level in is parsed from the log message. Currently, we are supporting following [LogLevels](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/logs.ts#L9-L24). So if you want Grafana to retrieve it, you just need to add required level to your log message. Maybe this file with [tests](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/utils/logs.test.ts#L12-L48) help. 🙂Also [here](https://play.grafana.org/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22InfluxDB%22,%7B%22resultFormat%22:%22table%22,%22groupBy%22:%5B%5D,%22select%22:%5B%5B%7B%22type%22:%22field%22,%22params%22:%5B%22description%22%5D%7D%5D%5D,%22tags%22:%5B%5D,%22limit%22:%221000%22,%22measurement%22:%22logs%22,%22policy%22:%22default%22,%22orderByTime%22:%22ASC%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) is the example of logs with `unknown` level and [here](https://user-images.githubusercontent.com/30407135/67639609-4088ec00-f8f2-11e9-82e5-b5236a555385.gif) is the gif of influx log where the level is retrieved.

**Comment by jcdauchy-moodys** [comment]
Thanks for the quick feedback, so if I understand correctly, if using the LogPanel, I need to add the level information in the message (what I had noticed by I found this a bit redundant.. Loki does it properly..). or I use TablePanel and I add the level in query (table format output).

Is it planned to add the way to add the level information (as done for Elasticsearch datasource) in the future ?

Thanks again

**Pull request merged**